File: kakao.app/js/ked/view/base/bubble.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.hasOwn(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}}

var bubble

import kxk from "../../../kxk.js"
let slash = kxk.slash
let kseg = kxk.kseg

import theme from "../../theme/theme.js"

import syntax from "../../util/syntax.js"

import view from "../base/view.js"


bubble = (function ()
{
    _k_.extend(bubble, view)
    function bubble (screen, name)
    {
        this["onMouse"] = this["onMouse"].bind(this)
        this["onMouseEnter"] = this["onMouseEnter"].bind(this)
        bubble.__super__.constructor.call(this,screen,name)
        this.pointerType = 'pointer'
        this.syntax = new syntax
        this.syntax.setExt('noon')
        this.rounded = ''
        this.setColor('bg',theme.hover.blur)
        this.setColor('hover',theme.hover.bg)
        this.setColor('empty',theme.editor.bg)
    }

    bubble.prototype["draw"] = function ()
    {
        var bg, ch, fg, x

        if (this.hidden())
        {
            return
        }
        bubble.__super__.draw.call(this)
        bg = (this.hover ? this.color.hover : this.color.bg)
        var list = _k_.list(this.rounded)
        for (x = 0; x < list.length; x++)
        {
            ch = list[x]
            if (_k_.in(ch,''))
            {
                this.cells.set(x,0,ch,bg,this.color.empty)
            }
            else
            {
                fg = this.syntax.getColor(x,0,ch)
                this.cells.set(x,0,ch,fg,bg)
                if (this.hover)
                {
                    this.cells.adjustContrastForHighlight(x,0,bg)
                }
            }
        }
    }

    bubble.prototype["set"] = function (item)
    {
        this.item = item
    
        var _62_28_, _62_42_

        if (_k_.empty(this.item))
        {
            return this.rounded = ''
        }
        else
        {
            ;(typeof this.syntax.setSegls === "function" ? this.syntax.setSegls(((_62_42_=this.item.segls) != null ? _62_42_ : [kseg(item.tilde)])) : undefined)
            return this.rounded = '' + this.item.tilde + ''
        }
    }

    bubble.prototype["onMouseEnter"] = function (event)
    {
        this.emit('action','enter')
        return bubble.__super__.onMouseEnter.call(this,event)
    }

    bubble.prototype["onMouse"] = function (event)
    {
        bubble.__super__.onMouse.call(this,event)
    
        switch (event.type)
        {
            case 'press':
                if (this.hover)
                {
                    this.emit('action','click',this.item)
                    return {redraw:true}
                }
                break
        }

        return this.hover
    }

    return bubble
})()

export default bubble;

File: kakao.app/love/lua/view/base/bubble.lua
--[[
    0000000    000   000  0000000    0000000    000      00000000  
    000   000  000   000  000   000  000   000  000      000       
    0000000    000   000  0000000    0000000    000      0000000   
    000   000  000   000  000   000  000   000  000      000       
    0000000     0000000   0000000    0000000    0000000  00000000  

    displays some clickable text in a rounded 'bubble'
    used by 
        ◆ searcher to display file headings
        ◆ status to display filepos status
--]]

syntax = require "util.syntax"
view = require "view.base.view"


local bubble = class("bubble", view)
    


function bubble:init(name) 
        view.init(self, name)
        
        self.pointerType = 'pointer'
        self.syntax = syntax()
        self.syntax:setExt('noon')
        
        self.rounded = array() -- ''
        
        self:setColor('bg', theme.hover.blur)
        self:setColor('hover', theme.hover.bg)
        self:setColor('empty', theme.editor.bg)
        return self
    end

-- 0000000    00000000    0000000   000   000  
-- 000   000  000   000  000   000  000 0 000  
-- 000   000  0000000    000000000  000000000  
-- 000   000  000   000  000   000  000   000  
-- 0000000    000   000  000   000  00     00  


function bubble:draw() 
        if self:hidden() then return end
        
        local bg = (function () 
    if self.hover then 
    return self.color.hover else 
    return self.color.bg
             end
end)()
        
        for x, ch in ipairs(self.rounded) do 
            if ((ch == '') or (ch == '')) then 
                self.cells:set(x, 1, ch, bg, self.color.empty)
            else 
                local fg = self.syntax:getColor(x, 0, ch)
                self.cells:set(x, 1, ch, fg, bg)
                if self.hover then 
                    self.cells:adjustContrastForHighlight(x, 1, bg)
                end
            end
        end
        
        return self:render()
    end


function bubble:set(item) 
        self.item = item
        
        if empty(self.item) then 
            self.rounded = array()
            return self.rounded
        else 
            self.syntax.setSegls((self.item.segls or array(kseg(item.tilde))))
            self.rounded = ((array('') + kseg(item.tilde)) + array(''))
            return self.rounded
        end
    end

-- 00     00   0000000   000   000   0000000  00000000  
-- 000   000  000   000  000   000  000       000       
-- 000000000  000   000  000   000  0000000   0000000   
-- 000 0 000  000   000  000   000       000  000       
-- 000   000   0000000    0000000   0000000   00000000  


function bubble:onMouseEnter(event) 
        self:emit('action', 'enter')
        return view.onMouseEnter(self, event)
    end


function bubble:onMouse(event) 
        view.onMouse(self, event)
        
        if (event.type == 'press') then 
                if self.hover then 
                    self:emit('action', 'click', self.item)
                    return {redraw = true}
                end
        end
        
        return self.hover
    end

return bubble